kytos/core/interface.py

```diff
@@ -173,19 +193,311 @@ def status_reason(self):
+        tag_n = len(tag_ranges)
+        used_n = len(used_tags)
+        tag_i, used_i = 0, 0
+        while tag_i < tag_n and used_i < used_n:
+            start_tag, end_tag = tag_ranges[tag_i]
+            start_used, end_used = used_tags[used_i]
+
+            if end_tag < start_used:
+                tag_i += 1
+            elif end_used < start_tag:
+                msg = f"Missing all tags from {used_tags[used_i]}"
+                raise KytosSetTagRangeError(msg)
+            else:
+                if start_tag <= start_used and end_tag >= end_used:
+                    tag_i += 1
+                    used_i += 1
+                else:
+                    msg = f"Missing tags from {used_tags[used_i]}"
+                    raise KytosSetTagRangeError(msg)
+
+        # Check if there is anything left
+        if used_i != used_n:
+            msg = f"Missing all tags from {used_tags[used_i:used_n]}"
+            raise KytosSetTagRangeError(msg)
+
+    def set_tag_ranges(self, tag_ranges: list[list[int]], tag_type: str):
+        """Set new restriction"""
+        if tag_type != TAGType.VLAN.value:
+            msg = f"Tag type {tag_type} is not supported."
+            raise KytosTagtypeNotSupported(msg)
+        with self._tag_lock:
+            used_tags = self.range_difference(
+                self.tag_ranges[tag_type], self.available_tags[tag_type]
+            )
+            # Verify new tag_ranges
+            try:
+                self.can_set_tag_ranges(used_tags, tag_ranges)
+            except KytosSetTagRangeError as err:
+                raise err
+            # Resizing
+            new_tag_ranges = self.range_difference(
+                tag_ranges, used_tags
+            )
+            self.available_tags[tag_type] = new_tag_ranges
+            self.tag_ranges[tag_type] = tag_ranges
+
+    def remove_tag_ranges(self, tag_type: str):
+        """Set tag_ranges[tag_type] to default [[1, 4095]]"""
+        if tag_type != TAGType.VLAN.value:
+            msg = f"Tag type {tag_type} is not supported."
+            raise KytosTagtypeNotSupported(msg)
+        with self._tag_lock:
+            used_tags = self.range_difference(
+                self.tag_ranges[tag_type], self.available_tags[tag_type]
+            )
+            self.available_tags[tag_type] = self.range_difference(
+                self.default_tag_values[tag_type], used_tags
+            )
+            self.tag_ranges[tag_type] = self.default_tag_values[tag_type]
+
+    @staticmethod
+    def find_index_remove(
+        available_tags: list[list[int]],
+        tag_range: list[int]
+    ) -> Optional[int]:
+        """Find the index of tags in available_tags to be removed"""
+        index = bisect.bisect_left(available_tags, tag_range)
+        if (index < len(available_tags) and
+                tag_range[0] >= available_tags[index][0] and
+                tag_range[1] <= available_tags[index][1]):
+            return index
+
+        if (index-1 > -1 and
```

Comment:
This is to remove a tag, so all the tags from `tag_range` should be inside a range from `available_tags`. When:
```
available_tags = [[10, 14], [20, 50], [100, 200]]
tag_range = [20, 50]
```
`bisect_left` will return 1 which is going to be examined by the first if statement and return 1 as a result. But when we have:
```
available_tags = [[10, 14], [20, 50], [100, 200]]
tag_range = [11, 13]
```
`bisect_left` will return 1 too. Comparing `available_tags[1] = [20, 50]` to `[11, 13]]` from the first if statement would result in None. In these cases, the index will have to be reduced by one `index = 0` and compare `available_tags[0] = [10, 14]` to `[11, 13]` which will result in the correct return 0.

Seeing that we have two possible cases, the index got from `bisect_left` is more like a suggestion. Since `index` can be 0 and the second if statement is necessary, `index-1>-1` exists to make sure that we do not look up for `available_tags[-1]`. Although it does not return an error but wanted to make sure that an unexpected behavior does not happen.

Reply:
This is to remove a tag, so all the tags from `tag_range` should be inside a range from `available_tags`. When:
```
available_tags = [[10, 14], [20, 50], [100, 200]]
tag_range = [20, 50]
```
`bisect_left` will return 1 which is going to be examined by the first if statement and return 1 as a result. But when we have:
```
available_tags = [[10, 14], [20, 50], [100, 200]]
tag_range = [11, 13]
```
`bisect_left` will return 1 too. Comparing `available_tags[1] = [20, 50]` to `[11, 13]]` from the first if statement would result in None. In these cases, the index will have to be reduced by one `index = 0` and compare `available_tags[0] = [10, 14]` to `[11, 13]` which will result in the correct return 0.

Seeing that we have two possible cases, the index got from `bisect_left` is more like a suggestion. Since `index` can be 0 and the second if statement is necessary, `index-1>-1` exists to make sure that we do not look up for `available_tags[-1]`. Although it does not return an error but wanted to make sure that an unexpected behavior does not happen.